### PR TITLE
WT-11230 Cleanup the usage of test_env_vars and configure_env_vars in evergreen.yml

### PR DESCRIPTION
**Pull request by jiechenbo**
The idea behind this change is that we utilize the concept of pre: tasks and expansions to help us clean up the environment variables. pre: will be run for every task in the evergreen system, and an expansion is a way to persist environment variables across tasks. In our case we have two expansions:

- PREPARE_TEST_ENV: Which allows the task to prepare any environment variables to run tests.
- PREPARE_PATH: Allows any environments that require the mongodbtoolchain path to be set for you.

The idea is that we can run the expansions at the start of each task, which is meant to do the setup for you. I have been able to clean up multiple different variables in the buildvariants through this.

### CODE REVIEW

File: test/evergreen.yml [hunk 1]
```diff
@@ -199,7 +230,7 @@ functions:
       script: |
         set -o errexit
         # Confirm that the Python binary matches the version of that configured by CMake.
-        ${python_binary} ../test/evergreen/python_version_check.py -v -c ./CMakeCache.txt -s ${python_config_search_string}
+        ${python_binary|python3} ../test/evergreen/python_version_check.py -v -c ./CMakeCache.txt -s ${python_config_search_string}
```
Comment (etienneptl) [inline test/evergreen.yml:233]: Do we still need `python_binary`? The pattern `${python_binary|python3}` is used multiple times in this file.

File: test/evergreen.yml [hunk 2]
```diff
@@ -208,13 +239,18 @@ functions:
...
         set -o verbose
+        ${PREPARE_PATH}
         if [ "$OS" = "Windows_NT" ]; then
           # Use the Windows powershell script to execute Ninja build (can't execute directly in a cygwin environment).
           powershell.exe '.\test\evergreen\build_windows.ps1 -build 1'
         else
           # Compiling with CMake generated Ninja file.
           cd cmake_build
-          ${compile_env_vars|} ${make_command|ninja} ${smp_command|} 2>&1
+          if [ "${cmake_generator|Ninja}" = "Ninja" ]; then
```
Comment (etienneptl) [inline test/evergreen.yml:249]: Where do we use something else than `Ninja`?

File: test/evergreen.yml [hunk 3]
```diff
@@ -1159,10 +1221,9 @@ variables:
       CC_OPTIMIZE_LEVEL: -DCC_OPTIMIZE_LEVEL=-O0
       smp_command: -j $(sysctl -n hw.logicalcpu)
       cmake_generator: "Unix Makefiles"
```
Comment (etienneptl) [inline test/evergreen.yml:1223]: I guess this line answers my question regarding `Ninja`. Do we know why we are using `Unix Makefiles` here? If this is because it is not available on macOS, fair enough, my comments can be resolved.

File: test/evergreen.yml [hunk 4]
```diff
@@ -28,6 +29,35 @@ exec_timeout_secs: 21600 # 6 hrs
...
+            else
+              export PATH=/opt/mongodbtoolchain/v4/bin:$PATH
+              export LD_LIBRARY_PATH=$WT_BUILDDIR
+            fi
+
+            if [[ "${CMAKE_PREFIX_PATH|}" =~ TCMALLOC_LIB ]]; then
+              export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$WT_TOPDIR/TCMALLOC_LIB/lib"
+              export LD_PRELOAD=$WT_TOPDIR/TCMALLOC_LIB/lib/libtcmalloc.so
+            fi
+            ${additional_env_vars}
```
Comment (etienneptl) [inline test/evergreen.yml:53]: This comes from the build variant definition, right? I don't know where is the right place to do so, but we should have a comment describing what `PREPARE_TEST_ENV` is for and when to use `additional_env_vars`. I believe we could simply say one is common to all build variants and the other is used when the build variant needs extra parameters.

File: test/evergreen_develop.yml [hunk 5]
```diff
@@ -178,19 +157,9 @@ buildvariants:
...
-    test_env_vars:
-      PATH=/opt/mongodbtoolchain/v4/bin:$PATH
-      WT_TOPDIR=$(git rev-parse --show-toplevel)
-      WT_BUILDDIR=$WT_TOPDIR/cmake_build
-      LD_LIBRARY_PATH=$WT_BUILDDIR:$WT_TOPDIR/TCMALLOC_LIB/lib
-      LD_PRELOAD=$WT_TOPDIR/TCMALLOC_LIB/lib/libtcmalloc.so
     CMAKE_PREFIX_PATH: -DCMAKE_PREFIX_PATH="$(pwd)/TCMALLOC_LIB"
-    python_binary: '/opt/mongodbtoolchain/v4/bin/python3'
     smp_command: -j $(echo "`grep -c ^processor /proc/cpuinfo` * 2" | bc)
     cmake_generator: Ninja
```
Comment (etienneptl) [inline test/evergreen_develop.yml:162]: Do we need to specify `cmake_generator`? It seems that the default value `Ninja` is used.

**Review by etienneptl — COMMENTED** [review]
I like those changes! Good job, I have left a few comments.

File: test/evergreen.yml (hunk 1 above)
Comment (jiechenbo) [reply test/evergreen.yml:233]: yes unfortunately some tasks require their own python_binary to be set. E.g. windows, or tasks that use their own virtual env.

File: test/evergreen_develop.yml (hunk 5 above)
Comment (jiechenbo) [reply test/evergreen_develop.yml:162]: I agree, do you think we should just default to Ninja instead? I can understand it can be used for readability to acknowledge a variant is using Ninja vs Make file. @lukech what do you think?

File: test/evergreen.yml (hunk 3 above)
Comment (jiechenbo) [reply test/evergreen.yml:1223]: I believe this is because the machine doesn't have ninja installed. @lukech  can you help confirm this?

File: test/evergreen.yml [hunk 6]
```diff
@@ -28,6 +29,39 @@ exec_timeout_secs: 21600 # 6 hrs
...
```
Comment (lukech) [inline test/evergreen.yml:56]: There seems to be inconsistency in the later setting points - some have `export` (making the expansion a command) others do not.

File: test/evergreen_develop.yml (hunk 5 above)
Comment (lukech) [reply test/evergreen_develop.yml:162]: I'm okay with `cmake_generator` setting line being removed (if the value is "Ninja" - the default).

File: test/evergreen.yml (hunk 3 above)
Comment (lukech) [reply test/evergreen.yml:1227]: I don't have a recollection of this, and can see the rhel80 and amazon2-arm64 builders also use make instead of ninja, which is unlikely about package missing  (though could be copy-paste).

Comment (lukech) [reply test/evergreen.yml:1227]: One possibility is to maintain the testing diversity to have both covered.

File: test/evergreen.yml [hunk 7]
```diff
@@ -5761,19 +5837,9 @@ buildvariants:
...
   expansions:
-    configure_env_vars: PATH=/opt/mongodbtoolchain/v4/bin:$PATH
-    compile_env_vars: PATH=/opt/mongodbtoolchain/v4/bin:$PATH
-    test_env_vars:
-      PATH=/opt/mongodbtoolchain/v4/bin:$PATH
-      WT_TOPDIR=$(git rev-parse --show-toplevel)
-      WT_BUILDDIR=$WT_TOPDIR/cmake_build
-      LD_LIBRARY_PATH=$WT_BUILDDIR:$WT_TOPDIR/TCMALLOC_LIB/lib
-      LD_PRELOAD=$WT_TOPDIR/TCMALLOC_LIB/lib/libtcmalloc.so
     CMAKE_PREFIX_PATH: -DCMAKE_PREFIX_PATH="$(pwd)/TCMALLOC_LIB"
```
Comment (lukech) [inline test/evergreen.yml:5840]: This line of setting is used for most builders. This could be another improvement opportunity to make the expansion adoption place (the "configure wiredtiger" func) set a default value. Happy for it to be covered in a separate PR/ticket.

File: test/evergreen.yml [hunk 8]
```diff
@@ -6044,16 +6057,7 @@ buildvariants:
   # We run on medium as small has too small a disk.
   - ubuntu2004-perf-testing
   expansions:
-    configure_env_vars: PATH=/opt/mongodbtoolchain/v4/bin:$PATH
```
Comment (lukech) [inline test/evergreen.yml:6047]: This looks like a pretty standard expansions setting (after simplification by this PR). Should we consider making some expansions template to further reduce duplication? (again, happy it to be covered in a separate PR/ticket)

File: test/evergreen.yml [hunk 9]
```diff
@@ -28,6 +29,39 @@ exec_timeout_secs: 21600 # 6 hrs
...
+        - key: PREPARE_TEST_ENV
+          value: |
+            export WT_TOPDIR=$(git rev-parse --show-toplevel)
+            export WT_BUILDDIR=$WT_TOPDIR/cmake_build
+
+            if [ "$OS" = "Windows_NT" ]; then
+              export PATH=/cygdrive/c/python/Python311:/cygdrive/c/python/Python311/Scripts:$PATH
+              export PYTHONPATH="($WT_TOPDIR/lang/python/wiredtiger):$(cygpath -w $WT_TOPDIR/lang/python)"
+            else
+              export PATH=/opt/mongodbtoolchain/v4/bin:$PATH
```
Comment (lukech) [inline test/evergreen.yml:48]: Please note neither Windows nor macOS has toolchain support. I don't expect any issue with setting PATH for their builders, but may be worth paying special attention during testing to make sure no surprises.

File: test/evergreen.yml [hunk 10]
```diff
@@ -28,6 +29,39 @@ exec_timeout_secs: 21600 # 6 hrs
 #######################################
 
 functions:
+  "setup environment":
+    - command: expansions.update
```
Comment (lukech) [inline test/evergreen.yml:33]: Thanks for introducing this Evergreen command to me. While it's not super intuitive for me to understand at which stage these expansion updates are applied, and the need of calling the update key (e.g. PREPARE_TEST_ENV) explicitly in the task definition, it could be a similar experience for other readers. A comment could be helpful.

File: test/evergreen_develop.yml (hunk 5 above)
Comment (jiechenbo) [reply test/evergreen_develop.yml:162]: [3856cab](https://github.com/wiredtiger/wiredtiger/pull/9533/commits/3856cab1ed5f29fae68eca15064a0b18d9609cfb)

File: test/evergreen.yml (hunk 6 above)
Comment (jiechenbo) [reply test/evergreen.yml:56]: fixed! [3856cab](https://github.com/wiredtiger/wiredtiger/pull/9533/commits/3856cab1ed5f29fae68eca15064a0b18d9609cfb)

File: test/evergreen.yml [hunk 11]
```diff
@@ -6044,16 +6057,7 @@ buildvariants:
...
-    compile_env_vars: PATH=/opt/mongodbtoolchain/v4/bin:$PATH
-    test_env_vars:
-      PATH=/opt/mongodbtoolchain/v4/bin:$PATH
-      WT_TOPDIR=$(git rev-parse --show-toplevel)
-      WT_BUILDDIR=$WT_TOPDIR/cmake_build
-      LD_LIBRARY_PATH=$WT_BUILDDIR:$WT_TOPDIR/TCMALLOC_LIB/lib
-      LD_PRELOAD=$WT_TOPDIR/TCMALLOC_LIB/lib/libtcmalloc.so
     cmake_generator: Ninja
-    make_command: ninja
     CMAKE_PREFIX_PATH: -DCMAKE_PREFIX_PATH="$(pwd)/TCMALLOC_LIB"
```
[the remark this answers is not in the document]
Comment (jiechenbo) [reply test/evergreen.yml:6061]: Yep! Created [WT-11530](https://jira.mongodb.org/browse/WT-11530)

File: test/evergreen.yml (hunk 7 above)
Comment (jiechenbo) [reply test/evergreen.yml:5840]: Yep! Created [WT-11530](https://jira.mongodb.org/browse/WT-11530)

File: test/evergreen.yml (hunk 10 above)
Comment (jiechenbo) [reply test/evergreen.yml:33]: done!

File: test/evergreen.yml (hunk 3 above)
Comment (jiechenbo) [reply test/evergreen.yml:1228]: For this ticket we should focus on cleaning up the existing variables rather than adding more tests. I am happy to create a ticket to investigate.

**Review by lukech — APPROVED** [review]
LGTM, thanks.

**Review by etienneptl — APPROVED** [review]
LGTM, thank you for all the effort!

**Comment by jiechenbo** [comment]
evergreen merge
